js/msg.js

```diff
@@ -1,258 +1,168 @@
+    return fn && fn();
   }
 
-  function initHandler() {
-    if (handler) {
-      return;
-    }
-    handler = {
-      both: [],
-      tab: [],
-      extension: []
-    };
-    if (isBg) {
-      window._msg.handler = handler;
-    }
-    chrome.runtime.onMessage.addListener(handleMessage);
+  function getExtBgIfReady() {
+    const bg = getExtBg();
+    return bg && bg.document && bg.document.readyState !== 'loading' && bg;
   }
 
-  function executeCallbacks(callbacks, ...args) {
-    let result;
-    for (const fn of callbacks) {
-      const data = withPromiseError(fn, ...args);
-      if (data !== undefined && result === undefined) {
-        result = data;
-      }
-    }
-    return result;
+  function getRuntimeBg() {
+    return new Promise(resolve =>
+      chrome.runtime.getBackgroundPage(bg =>
+        resolve(!chrome.runtime.lastError && bg)));
   }
 
-  function handleMessage(message, sender, sendResponse) {
-    const handlers = message.target === 'tab' ?
-      handler.tab.concat(handler.both) : message.target === 'extension' ?
-      handler.extension.concat(handler.both) :
-      handler.both.concat(handler.extension, handler.tab);
-    if (!handlers.length) {
-      return;
+  function handleMessage({data, target}, sender, sendResponse) {
+    const res = msg._execute(TARGETS[target] || TARGETS.all, data, sender);
+    if (res instanceof Promise) {
+      handleResponseAsync(res, sendResponse);
+      return true;
     }
-    const result = executeCallbacks(handlers, message.data, sender);
-    if (result === undefined) {
-      return;
-    }
-    Promise.resolve(result)
-      .then(
-        data => ({
-          error: false,
-          data
-        }),
-        err => ({
-          error: true,
-          data: Object.assign({
-            message: err.message || String(err),
-            // FIXME: do we want to pass the entire stack?
-            stack: err.stack
-          }, err) // this allows us to pass custom properties e.g. `err.index`
-        })
-      )
-      .then(sendResponse);
-    return true;
+    if (res !== undefined) sendResponse({data: res});
   }
 
-  function withPromiseError(fn, ...args) {
+  async function handleResponseAsync(promise, sendResponse) {
     try {
-      return fn(...args);
+      sendResponse({
+        data: await promise,
+      });
     } catch (err) {
-      return Promise.reject(err);
+      sendResponse({
+        error: true,
+        data: Object.assign({
+          message: err.message || String(err),
+          stack: err.stack,
+        }, err), // passing custom properties e.g. `err.index` to unwrapData
+      });
     }
   }
 
-  // {type, error, data, id}
-  function unwrapData(result) {
-    if (result === undefined) {
-      throw new Error('Receiving end does not exist');
-    }
-    if (result.error) {
-      throw Object.assign(new Error(result.data.message), result.data);
-    }
-    return result.data;
+  function unwrapData({data, error} = {}) {
```

Comment:
This won't throw when the remote handler doesn't exist. It should probably be:
```js
({data, error} = {
  data: {message: 'Receiving end does not exist'},
  error: true
})
```

BTW, maybe we should move the error handling stuff into `polyfill.js` in the future. In Firefox, you can return a promise in `addListener` callback. If it fails, the error object will be passed to the caller correctly. Therefor `msg.js` will only include `sendBg` and the `API` stuff.